Building blocks (#5)
Troblecodings/TGDocument: markdown/05_building_block_view.md

# Building Block View

## ShaderTool

> The primary purpose of the ShaderTool was to compile shader into the Engine. The tool started to grow because of it's console interface which makes it ideal for tooling.

Reviewer: ```suggestion
The primary purpose of the ShaderTool was to compile shader into the Engine. The tool started to grow because of its console interface, which makes it ideal for tooling.
```

![Whitebox](images/whitebox1.png)

| Subsystem         | Description                                                                  |
| ----------------- | ---------------------------------------------------------------------------- |

> | Console Interface | Allowes for Console input and forwards the arguments to the give subsystem   |

Reviewer: ```suggestion
| Console Interface | Allow s for console input and forwards the arguments to the given subsystem   |
```

Reviewer: whoops don't mind the space in "allows"

> | Shader Interface  | This system compiles shader and embeds them into the Engine                  |

Reviewer: ```suggestion
| Shader Interface  | This system compiles shader and embeds them into the engine                  |
```

| Pipe Interface    | Analyzes the compiled shaders and creates the according pipeline information |

> | Ressource System  | This system administrates all game ressources.                               |

Reviewer: ```suggestion
| Resource System   | This system administrates all game resources.                                |
```

## TGEngine

> The decomposition has historical resons

Reviewer: ```suggestion
The decomposition has historical reasons
```

| --------------- | ---------------------------------------------------------------------------------------------------------------------------------- |
| IO              | For general in and output managment                                                                                                |
| UI              | Contains all user interface components                                                                                             |
| Gamecontent     | Everything that can be put into the game (Actors, Lights  ...)                                                                     |
| Pipeline        | This is the heart of the engine and contains everything needed to create a Vulkan Application with the according featureset needed |
| General IO      | Platform independency layer for file handling                                                                                      |
| Properties      | A small XML reader with a given schema                                                                                             |

> | Ressourcesystem | the counterpart to the ShaderTools Ressource System, except this one only reads the ressources                                     |

Reviewer: ```suggestion
| Resourcesystem  | the counterpart to the ShaderTools resource system, except this one only reads the resources                                      
 |
```

> | Memory Layer    | The own Memory managmet for the GPU as well as the system memory                                                                   |

Reviewer: ```suggestion
| Memory Layer    | The own memory management for the GPU, as well as the system memory                                                                |
```
